### ISSUE TITLE
ETS table name conflict racing if it is created in a supervised gen_server:init

### ISSUE THREAD

**Issue by zzydxm** [erlang/otp#6339]
**Describe the bug**
In a module which implements gen_server behavior, if we create an named ETS table in the init callback, and the gen_server crashed, then the restart might not be successful due to table name conflict, because the table is not freed yet. e.g.
```
    supervisor: {<0.87.0>,sup}
    errorContext: start_error
    reason: {badarg,[{ets,new,[a,[named_table]],...
```

The root cause of this problem is that ```gen_server:start_link``` can return ```{error, Reason}``` before the process dies (https://github.com/erlang/otp/blob/master/lib/stdlib/src/gen_server.erl#L846-L847 There is still an ```erlang:raise``` operation left), so the ETS table is still alive. Then supervisor will do a restart of the child and call ```ets:new``` again with the same name and fail.

One potential solution is to add a ```proc_lib:init_ack(Parent, Child, Return)``` function, that spawns a function which monitors the ```Child```, wait until receive ```'DOWN'``` message, and then send ack to ```Parent```.

I know this problem might have been discussed before, but would like to hear about some context on whether there is a plan to fix it or why not to fix it. Thanks!

**To Reproduce**
sup.erl:
```
-module(sup).
-export([init/1]).
init([]) ->
    {ok, {#{intensity => 1000}, [#{id => a, start => {gen_server, start_link, [sup, a, []]}}]}};
init(a) ->
    ets:new(a, [named_table]),
    timer:sleep(10),
    persistent_term:get(a, false) andalso (a = b),
    persistent_term:put(a, true),
    self() ! {'$gen_cast', a},
    {ok, #{}}.
```
terminal:
```
erlc sup.erl; erl
supervisor:start_link(sup, []).
```
error:
```
=SUPERVISOR REPORT==== 29-Sep-2022::15:29:21.434913 ===
    supervisor: {<0.87.0>,sup}
    errorContext: start_error
    reason: {badarg,[{ets,new,
                          [a,[named_table]],
                          [{error_info,#{module => erl_stdlib_errors}}]},
                     {sup,init,1,[{file,"sup.erl"},{line,6}]},
                     {gen_server,init_it,2,
                                 [{file,"gen_server.erl"},{line,851}]},
                     {gen_server,init_it,6,
                                 [{file,"gen_server.erl"},{line,814}]},
                     {proc_lib,init_p_do_apply,3,
                               [{file,"proc_lib.erl"},{line,240}]}]}
    offender: [{pid,{restarting,<0.88.0>}},
               {id,a},
               {mfargs,{gen_server,start_link,[sup,a,[]]}},
               {restart_type,permanent},
               {significant,false},
               {shutdown,5000},
               {child_type,worker}]

=CRASH REPORT==== 29-Sep-2022::15:29:21.435013 ===
  crasher:
    initial call: sup:init/1
    pid: <0.177.0>
    registered_name: []
    exception error: bad argument
      in function  ets:new/2
         called as ets:new(a,[named_table])
         *** argument 2: invalid options
      in call from sup:init/1 (sup.erl, line 6)
      in call from gen_server:init_it/2 (gen_server.erl, line 851)
      in call from gen_server:init_it/6 (gen_server.erl, line 814)
    ancestors: [<0.87.0>,<0.85.0>]
    message_queue_len: 0
    messages: []
    links: [<0.87.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 376
    stack_size: 28
    reductions: 156
  neighbours:
```

**Expected behavior**
```gen_server``` pid should die before returning error on start. ```ets:new``` should always success in this case

**Affected versions**
25.0.2

**Comment by josevalim** [comment]
> One potential solution is to add a proc_lib:init_ack(Parent, Child, Return) function, that spawns a function which monitors the Child, wait until receive 'DOWN' message, and then send ack to Parent.

Another implementation, based on the same API, is to return the Child to the parent. Then the Parent will immediately monitor the Child and return Return only when the Child dies. The benefit is that it does not take an additional process nor additional copying of `Return`.

**Comment by RaimoNiskanen** [comment]
I am for @josevalim's idea.  Kind of.  Introduce a `proc_lib:init_nack/1,2` to use in `gen_`* when the process init fails.  `proc_lib` will then await a monitor down or exit signal before returning.

We could _almost_ add this to `proc_lib:init_ack/1,2` by looking at the result value; `{ok,Pid}` is succesful start, and `ignore | {error, _}` is failure.  But since `proc_lib` can be used for a server that does not follow that convention it is safer to add a new API function to use when the server did not start.  `proc_lib:sync_init_ack/1,2` is a different alternative for the name.

**Comment by RaimoNiskanen** [comment]
I have started sketching on a PR. Stay tuned...

**Comment by RaimoNiskanen** [comment]
We can discuss my suggested solution in the PR: #6843

**Comment by RaimoNiskanen** [comment]
A fix has been merged into the 'master' branch for OTP-26.0.  It had to be a potentially incompatible extension of `proc_lib` so it was not suitable for OTP-25.*.

It should be possible to, as a workaround, use `proc_lib:start_monitor/3,4,5` and wait for the `'DOWN'` message if there is an error return.

Thank you for reporting this issue!
